module/apmchi/middleware.go

```diff
@@ -0,0 +1,104 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package apmchi
+
+import (
+	"net/http"
+
+	"github.com/go-chi/chi"
+
+	"go.elastic.co/apm"
+	"go.elastic.co/apm/module/apmhttp"
+)
+
+// Middleware returns a new chi middleware handler
+// for tracing requests and reporting errors.
+//
+// By default, the middleware will use apm.DefaultTracer.
+// Use WithTracer to specify an alternative tracer.
+func Middleware(o ...Option) func(http.Handler) http.Handler {
+	opts := options{
+		tracer:         apm.DefaultTracer,
+		requestIgnorer: apmhttp.DefaultServerRequestIgnorer(),
+	}
+	for _, o := range o {
+		o(&opts)
+	}
+	return func(h http.Handler) http.Handler {
+		return apmhttp.Wrap(
+			h,
+			apmhttp.WithTracer(opts.tracer),
+			apmhttp.WithServerRequestName(routeRequestName),
+			apmhttp.WithServerRequestIgnorer(opts.requestIgnorer),
+		)
+	}
+}
+
+func routeRequestName(r *http.Request) string {
+	return r.Method + " " + getRequestPattern(r)
+}
+
+func getRequestPattern(r *http.Request) string {
+	rctx := chi.RouteContext(r.Context())
+	if pattern := rctx.RoutePattern(); pattern != "" {
+		return pattern
+	}
+
+	routePath := r.URL.Path
+	if r.URL.RawPath != "" {
+		routePath = r.URL.RawPath
+	}
+
+	tctx := chi.NewRouteContext()
+	if !rctx.Routes.Match(tctx, r.Method, routePath) {
+		return routePath
```

Comment:
To match the behaviour of other agents, I intend to change the Go agent to set the transaction name to `"<METHOD> unknown route"` when the route cannot be determined (https://github.com/elastic/apm-agent-go/issues/486), such as in this case. Would you mind making that change here?

Reply:
Done and tested.